File: cpp/ascii_video_player.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <stdexcept>
#include <thread>
#include <vector>

#include <opencv2/opencv.hpp>

namespace {
const std::string ASCII_GRADIENT = " .:-=+*#%@";
const std::string HIGH_CONTRAST_GRADIENT =
    " .'`^\",:;Il!i><~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$";

struct Options {
    std::string file;
    std::string mode = "enhanced-rgb";
    double fps = 30.0;
    int width = 120;
    int height = 40;
    bool optimize = false;
    bool invert = false;
    bool noColor = false;
    int startFrame = 0;
    int maxFrames = 0;  // 0 = all
    double downscale = 1.0;
    std::string charset = ASCII_GRADIENT;
    bool showInfo = false;
};

Options parseArgs(int argc, char** argv) {
    Options opt;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        auto nextValue = [&](int& target) {
            if (i + 1 < argc)
                target = std::stoi(argv[++i]);
        };
        auto nextDouble = [&](double& target) {
            if (i + 1 < argc)
                target = std::stod(argv[++i]);
        };
        auto nextString = [&](std::string& target) {
            if (i + 1 < argc)
                target = argv[++i];
        };

        if (arg == "--file") {
            nextString(opt.file);
        } else if (arg == "--mode") {
            nextString(opt.mode);
        } else if (arg == "--fps") {
            nextDouble(opt.fps);
        } else if (arg == "--width") {
            nextValue(opt.width);
        } else if (arg == "--height") {
            nextValue(opt.height);
        } else if (arg == "--optimize") {
            opt.optimize = true;
        } else if (arg == "--invert") {
            opt.invert = true;
        } else if (arg == "--no-color") {
            opt.noColor = true;
        } else if (arg == "--start") {
            nextValue(opt.startFrame);
        } else if (arg == "--frames") {
            nextValue(opt.maxFrames);
        } else if (arg == "--downscale") {
            nextDouble(opt.downscale);
        } else if (arg == "--charset") {
            nextString(opt.charset);
        } else if (arg == "--show-info") {
            opt.showInfo = true;
        }
    }
    return opt;
}

std::string pickGradient(const Options& opt) {
    if (opt.mode == "enhanced-bw") {
        return HIGH_CONTRAST_GRADIENT;
    }
    return opt.charset;
}

cv::Mat quantizeColor(const cv::Mat& frame, const std::string& mode) {
    cv::Mat output;
    if (mode == "mid-rgb") {
        frame.convertTo(output, CV_8UC3);
        output = output / 32 * 32;
    } else if (mode == "low-rgb") {
        frame.convertTo(output, CV_8UC3);
        output = output / 64 * 64;
    } else {
        output = frame.clone();
    }
    return output;
}

std::string frameToAscii(const cv::Mat& frame, const Options& opt) {
    const int targetWidth = std::max(4, static_cast<int>(opt.width / opt.downscale));
    const int targetHeight = std::max(2, static_cast<int>(opt.height / opt.downscale));

    cv::Mat resized;
    cv::resize(frame, resized, cv::Size(targetWidth, targetHeight), 0, 0, cv::INTER_AREA);

    cv::Mat gray;
    cv::cvtColor(resized, gray, cv::COLOR_BGR2GRAY);
    if (opt.mode == "enhanced-bw") {
        cv::equalizeHist(gray, gray);
    }

    const std::string gradient = pickGradient(opt);
    const int gradLen = static_cast<int>(gradient.size());

    std::ostringstream oss;

    auto pushAnsi = [&](int r, int g, int b, char ch) {
        oss << "\033[38;2;" << r << ';' << g << ';' << b << 'm' << ch;
    };

    cv::Mat colors = opt.noColor || opt.mode == "bw" || opt.mode == "enhanced-bw"
                         ? cv::Mat()
                         : quantizeColor(resized, opt.mode);

    for (int y = 0; y < gray.rows; ++y) {
        const uchar* row = gray.ptr<uchar>(y);
        const cv::Vec3b* colorRow = colors.empty() ? nullptr : colors.ptr<cv::Vec3b>(y);
        for (int x = 0; x < gray.cols; ++x) {
            float intensity = static_cast<float>(row[x]) / 255.0f;
            if (opt.invert) {
                intensity = 1.0f - intensity;
            }
            int idx = static_cast<int>(intensity * (gradLen - 1));
            idx = std::clamp(idx, 0, gradLen - 1);
            char ch = gradient[idx];
            if (colorRow) {
                const cv::Vec3b& pixel = colorRow[x];
                pushAnsi(pixel[2], pixel[1], pixel[0], ch);
            } else {
                oss << ch;
            }
        }
        if (colorRow) {
            oss << "\033[0m";
        }
        if (y + 1 < gray.rows) {
            oss << '\n';
        }
    }
    return oss.str();
}

std::vector<cv::Mat> loadFrames(cv::VideoCapture& cap, int startFrame) {
    std::vector<cv::Mat> frames;
    cap.set(cv::CAP_PROP_POS_FRAMES, startFrame);
    cv::Mat frame;
    while (cap.read(frame)) {
        frames.push_back(frame.clone());
    }
    return frames;
}

void render(const Options& opt) {
    if (opt.file.empty()) {
        throw std::runtime_error("--file is required");
    }

    cv::VideoCapture cap(opt.file);
    if (!cap.isOpened()) {
        throw std::runtime_error("Could not open video file: " + opt.file);
    }

    const int totalFrames = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_COUNT));
    const double targetInterval = 1.0 / std::max(opt.fps, 0.001);

    std::vector<cv::Mat> cached;
    if (opt.optimize) {
        cached = loadFrames(cap, opt.startFrame);
    }

    auto iteratorFrames = [&](auto&& callback) {
        int processed = 0;
        if (!cached.empty()) {
            for (const auto& frame : cached) {
                callback(frame, ++processed);
                if (opt.maxFrames > 0 && processed >= opt.maxFrames)
                    return;
            }
            return;
        }

        cap.set(cv::CAP_PROP_POS_FRAMES, opt.startFrame);
        cv::Mat frame;
        while (cap.read(frame)) {
            callback(frame, ++processed);
            if (opt.maxFrames > 0 && processed >= opt.maxFrames)
                break;
        }
    };

    auto start = std::chrono::steady_clock::now();

    iteratorFrames([&](const cv::Mat& frame, int index) {
        auto frameStart = std::chrono::steady_clock::now();

        std::string ascii = frameToAscii(frame, opt);
        std::cout << "\033[H\033[J" << ascii;
        if (opt.showInfo) {
            auto elapsed = std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count();
            double fps = index / std::max(0.0001, elapsed);
            std::cout << "\n[✓] Frame: " << index << " / " << (opt.maxFrames > 0 ? opt.maxFrames : totalFrames)
                      << " | FPS: " << std::fixed << std::setprecision(1) << fps
                      << " | Elapsed: " << std::setprecision(2) << elapsed << "s\n";
        }
        std::cout.flush();

        auto spent = std::chrono::duration<double>(std::chrono::steady_clock::now() - frameStart).count();
        double delay = targetInterval - spent;
        if (delay > 0) {
            std::this_thread::sleep_for(std::chrono::duration<double>(delay));
        }
    });
}
}  // namespace

int main(int argc, char** argv) {
    try {
        Options opt = parseArgs(argc, argv);
        render(opt);
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << "\n";
        return 1;
    }
    return 0;
}

File: ascii_video_player.py
import argparse
import sys
import time
from typing import Iterable, List, Sequence

import cv2
import numpy as np
from colorama import Style, init as colorama_init

# Ensure ANSI escape codes are reset properly on exit
colorama_init()

ASCII_GRADIENT = " .:-=+*#%@"
HIGH_CONTRAST_GRADIENT = " .'`^\",:;Il!i><~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$"


class FrameCache:
    def __init__(self, optimize: bool):
        self.optimize = optimize
        self.frames: List[np.ndarray] = []

    def load(self, capture: cv2.VideoCapture) -> None:
        if not self.optimize:
            return
        success, frame = capture.read()
        while success:
            self.frames.append(frame)
            success, frame = capture.read()

    def __iter__(self) -> Iterable[np.ndarray]:
        return iter(self.frames)

    def __len__(self) -> int:  # pragma: no cover - convenience
        return len(self.frames)


def parse_args(argv: Sequence[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Render a video as colorful ASCII in the terminal.")
    parser.add_argument("--file", required=True, help="Path to the input video file")
    parser.add_argument("--mode", default="enhanced-rgb", choices=[
        "enhanced-rgb",
        "mid-rgb",
        "low-rgb",
        "bw",
        "enhanced-bw",
    ], help="Rendering mode")
    parser.add_argument("--fps", type=float, default=30.0, help="Target playback FPS")
    parser.add_argument("--width", type=int, default=120, help="Frame width in characters")
    parser.add_argument("--height", type=int, default=40, help="Frame height in characters")
    parser.add_argument("--optimize", action="store_true", help="Pre-cache frames for smoother playback")
    parser.add_argument("--charset", default=ASCII_GRADIENT, help="Custom characters to use for shading")
    parser.add_argument("--invert", action="store_true", help="Invert the intensity mapping")
    parser.add_argument("--start", type=int, default=0, help="Start rendering from this frame number")
    parser.add_argument("--frames", type=int, default=0, help="Limit playback to this many frames (0 = all)")
    parser.add_argument("--no-color", action="store_true", help="Force grayscale even in RGB modes")
    parser.add_argument("--downscale", type=float, default=1.0, help="Additional downscale factor for performance")
    parser.add_argument("--show-info", action="store_true", help="Show FPS counter and progress")
    return parser.parse_args(argv)


def quantize_color(frame: np.ndarray, mode: str) -> np.ndarray:
    if mode == "enhanced-rgb":
        return frame
    if mode == "mid-rgb":
        return (frame // 32) * 32
    if mode == "low-rgb":
        return (frame // 64) * 64
    return frame


def intensity_map(gray: np.ndarray, charset: str, invert: bool) -> np.ndarray:
    gradient = charset
    indices = (gray.astype(np.float32) / 255.0) * (len(gradient) - 1)
    if invert:
        indices = (len(gradient) - 1) - indices
    return indices.astype(np.int16)


def frame_to_ascii(frame: np.ndarray, mode: str, width: int, height: int, charset: str, invert: bool, use_color: bool) -> str:
    # Downscale first for speed
    resized = cv2.resize(frame, (width, height), interpolation=cv2.INTER_AREA)
    gray = cv2.cvtColor(resized, cv2.COLOR_BGR2GRAY)
    if mode == "enhanced-bw":
        gray = cv2.equalizeHist(gray)
        gradient = HIGH_CONTRAST_GRADIENT
    else:
        gradient = charset

    idx = intensity_map(gray, gradient, invert)
    chars = np.array(list(gradient))[idx]

    if not use_color or mode in {"bw", "enhanced-bw"}:
        lines = ["".join(row.tolist()) for row in chars]
        return "\n".join(lines)

    colored = quantize_color(resized, mode)
    flat_chars = chars.reshape(-1)
    flat_colors = colored.reshape(-1, 3)
    fragments: List[str] = []
    for (b, g, r), ch in zip(flat_colors, flat_chars):
        fragments.append(f"\033[38;2;{r};{g};{b}m{ch}")
    lines = []
    for y in range(height):
        start = y * width
        end = start + width
        lines.append("".join(fragments[start:end]) + Style.RESET_ALL)
    return "\n".join(lines)


def render_video(options: argparse.Namespace) -> None:
    capture = cv2.VideoCapture(options.file)
    if not capture.isOpened():
        raise SystemExit(f"Could not open video file: {options.file}")

    total_frames = int(capture.get(cv2.CAP_PROP_FRAME_COUNT))
    target_interval = 1.0 / max(options.fps, 1e-3)
    charset = options.charset

    cache = FrameCache(options.optimize)
    if options.optimize:
        capture.set(cv2.CAP_PROP_POS_FRAMES, options.start)
        cache.load(capture)
        frames_iter: Iterable[np.ndarray] = cache
    else:
        capture.set(cv2.CAP_PROP_POS_FRAMES, options.start)
        frames_iter = iter(lambda: capture.read()[1] if capture.isOpened() else None, None)

    processed = 0
    start_time = time.perf_counter()
    for frame in frames_iter:
        if frame is None:
            break
        processed += 1
        if options.frames and processed > options.frames:
            break

        ascii_frame = frame_to_ascii(
            frame,
            mode=options.mode,
            width=max(4, int(options.width / options.downscale)),
            height=max(2, int(options.height / options.downscale)),
            charset=charset,
            invert=options.invert,
            use_color=not options.no_color,
        )

        frame_start = time.perf_counter()
        sys.stdout.write("\033[H\033[J")
        sys.stdout.write(ascii_frame)
        if options.show_info:
            elapsed = time.perf_counter() - start_time
            fps = processed / elapsed if elapsed > 0 else 0.0
            sys.stdout.write(
                f"\n[✓] Frame: {processed} / {options.frames or total_frames} | FPS: {fps:5.1f} | Elapsed: {elapsed:6.2f}s\n"
            )
        sys.stdout.flush()

        spent = time.perf_counter() - frame_start
        delay = target_interval - spent
        if delay > 0:
            time.sleep(delay)

    capture.release()


if __name__ == "__main__":
    opts = parse_args(sys.argv[1:])
    render_video(opts)
